File: controllers/university.controller.js
/**
 * Module dependencies.
 */
const models = require("../models");

/**
 * Display a listing of the resource.
 *
 * @param  Request  req
 * @param  Response  res
 * @param  Next  next
 *
 * @return Array
 */
exports.index = async (req, res, next) => {
  try {
    await models.university
      .findAll({
        attributes: {
          exclude: ["createdAt", "updatedAt"],
        },
      })
      .then((universities) => {
        res.status(202).json({
          status: "success",
          message: "Universities fetched successfully",
          data: universities || {},
        });
      })
      .catch((error) => {
        res.status(500).json({
          status: "error",
          message: "Internal server error",
          data: error.message || {},
        });
      });
  } catch (error) {
    res.status(500).json({
      status: "error",
      message: "Internal server error",
      data: error.message || {},
    });
  }
};

/**
 * Store a newly created resource in storage.
 *
 * @param  Request  req
 * @param  Response  res
 * @param  Next  next
 *
 * @return Array
 */
exports.store = async (req, res, next) => {
  const { name, description, major, quantity } = req.body;

  if (!name || !description || !major || !quantity) {
    return res.status(400).json({
      status: "error",
      message: "Bad request",
      data: {},
    });
  }

  try {
    await models.university
      .create({
        name,
        description,
        major,
        quantity,
      })
      .then((university) => {
        res.status(201).json({
          status: "success",
          message: "University created successfully",
          data: university || {},
        });
      })
      .catch((error) => {
        res.status(500).json({
          status: "error",
          message: "Internal server error",
          data: error.message || {},
        });
      });
  } catch (error) {
    res.status(500).json({
      status: "error",
      message: "Internal server error",
      data: error.message || {},
    });
  }
};

/**
 * Display the specified resource.
 *
 * @param  Request  req
 * @param  Response  res
 * @param  Next  next
 *
 * @return Array
 */
exports.show = async (req, res, next) => {
  const id = req.params.id;

  try {
    await models.university
      .findByPk(id, {
        rejectOnEmpty: true,
        attributes: {
          exclude: ["createdAt", "updatedAt"],
        },
      })
      .then((university) => {
        res.status(206).json({
          status: "success",
          message: "University fetched successfully",
          data: university || {},
        });
      })
      .catch((error) => {
        res.status(500).json({
          status: "error",
          message: "Internal server error",
          data: error.message || {},
        });
      });
  } catch (error) {
    res.status(500).json({
      status: "error",
      message: "Internal server error",
      data: error.message || {},
    });
  }
};

/**
 * Update the specified resource in storage.
 *
 * @param  Request  req
 * @param  Response  res
 * @param  Next  next
 *
 * @return Array
 */
exports.update = async (req, res, next) => {
  const id = req.params.id;
  const { name, description, major, quantity } = req.body;

  if (!name || !description || !major || !quantity) {
    return res.status(400).json({
      status: "error",
      message: "Bad request",
      data: {},
    });
  }

  try {
    await models.university
      .findByPk(id, {
        rejectOnEmpty: true,
        attributes: {
          exclude: ["createdAt", "updatedAt"],
        },
      })
      .then((university) => {
        university.name = name;
        university.description = description;
        university.major = major;
        university.quantity = quantity;

        return university.save();
      })
      .then((university) => {
        res.status(202).json({
          status: "success",
          message: "University updated successfully",
          data: university || {},
        });
      })
      .catch((error) => {
        res.status(500).json({
          status: "error",
          message: "Internal server error",
          data: error.message || {},
        });
      });
  } catch (error) {
    res.status(500).json({
      status: "error",
      message: "Internal server error",
      data: error.message || {},
    });
  }
};

/**
 * Remove the specified resource from storage.
 *
 * @param  Request  req
 * @param  Response  res
 * @param  Next  next
 *
 * @return Array
 */
exports.destroy = async (req, res, next) => {
  const id = req.params.id;

  try {
    await models.university
      .findByPk(id, { rejectOnEmpty: true })
      .then((university) => {
        return university.destroy();
      })
      .then(() => {
        res.status(202).json({
          status: "success",
          message: "University deleted successfully",
          data: null,
        });
      })
      .catch((error) => {
        res.status(500).json({
          status: "error",
          message: "Internal server error",
          data: error.message || {},
        });
      });
  } catch (error) {
    res.status(500).json({
      status: "error",
      message: "Internal server error",
      data: error.message || {},
    });
  }
};

File: src/controllers/university_controller.rs
use validator::Validate;
use serde_json::{Value, json};
use actix_web::{web, Responder};

use crate::{
  schemas::university_schema::*,
  structs::main_struct::AppState,
  services::university_service::*,
  helpers::response::create_response,
  enums::{
    error_enum::ErrorEnum,
    response_enum::ResponseDataEnum
  }
};

#[doc = "Display a listing of the resource."]
pub async fn university_index_controller(state: web::Data<AppState>) -> impl Responder {
  match university_index_service(state.db.to_owned()).await {
    Ok(data) => {
      create_response(
        String::from("success"),
        String::from("successfully retrieved university data"),
        data
      )
    },
    Err(err) => {
      match err {
        ErrorEnum::CustomError(message) => {
          create_response(
            String::from("unprocessable entity"),
            message,
            ResponseDataEnum::SingleValue(json!({}))
          )
        },
        _ => {
          create_response(
            String::from("internal server error"),
            err.get_error(),
            ResponseDataEnum::SingleValue(json!({}))
          )
        }
      }
    }
  }
}

#[doc = "Store a newly created resource in storage."]
pub async fn university_store_controller(state: web::Data<AppState>, body: web::Json<UniversitySchema>) -> impl Responder {
  let validate_form = body.validate();

  if validate_form.is_err() {
    let data = Value::from(validate_form.err().unwrap().to_string());

    return create_response(
      String::from("unprocessable entity"),
      String::from("please fill all fields"),
      ResponseDataEnum::SingleValue(data)
    )
  }

  match university_store_service(state.db.to_owned(), body.into_inner()).await {
    Ok(data) => {
      create_response(
        String::from("success"),
        String::from("successfully created university data"),
        data
      )
    },
    Err(err) => {
      match err {
        ErrorEnum::CustomError(message) => {
          create_response(
            String::from("unprocessable entity"),
            message,
            ResponseDataEnum::SingleValue(json!({}))
          )
        },
        _ => {
          create_response(
            String::from("internal server error"),
            err.get_error(),
            ResponseDataEnum::SingleValue(json!({}))
          )
        }
      }
    }
  }
}

#[doc = "Display the specified resource."]
pub async fn university_show_controller(state: web::Data<AppState>, path: web::Path<i32>) -> impl Responder {
  match university_show_service(state.db.to_owned(), path.into_inner()).await {
    Ok(data) => {
      create_response(
        String::from("success"),
        String::from("successfully retrieved university data"),
        data
      )
    },
    Err(err) => {
      match err {
        ErrorEnum::CustomError(message) => {
          create_response(
            String::from("unprocessable entity"),
            message,
            ResponseDataEnum::SingleValue(json!({}))
          )
        },
        _ => {
          create_response(
            String::from("internal server error"),
            err.get_error(),
            ResponseDataEnum::SingleValue(json!({}))
          )
        }
      }
    }
  }
}

#[doc = "Update the specified resource in storage."]
pub async fn university_update_controller(state: web::Data<AppState>, body: web::Json<UniversitySchema>, path: web::Path<i32>) -> impl Responder {
  let id = path.into_inner();
  let validate_form = body.validate();

  if validate_form.is_err() {
    let data = Value::from(validate_form.err().unwrap().to_string());

    return create_response(
      String::from("unprocessable entity"),
      String::from("please fill all fields"),
      ResponseDataEnum::SingleValue(data)
    )
  }

  match university_update_service(state.db.to_owned(), id, body.into_inner()).await {
    Ok(data) => {
      create_response(
        String::from("success"),
        String::from("successfully updated university data"),
        data
      )
    },
    Err(err) => {
      match err {
        ErrorEnum::CustomError(message) => {
          create_response(
            String::from("unprocessable entity"),
            message,
            ResponseDataEnum::SingleValue(json!({}))
          )
        },
        _ => {
          create_response(
            String::from("internal server error"),
            err.get_error(),
            ResponseDataEnum::SingleValue(json!({}))
          )
        }
      }
    }
  }
}

#[doc = "Remove the specified resource from storage."]
pub async fn university_destroy_controller(state: web::Data<AppState>, path: web::Path<i32>) -> impl Responder {
  match university_destroy_service(state.db.to_owned(), path.into_inner()).await {
    Ok(data) => {
      create_response(
        String::from("success"),
        String::from("successfully deleted university data"),
        data
      )
    },
    Err(err) => {
      match err {
        ErrorEnum::CustomError(message) => {
          create_response(
            String::from("unprocessable entity"),
            message,
            ResponseDataEnum::SingleValue(json!({}))
          )
        },
        _ => {
          create_response(
            String::from("internal server error"),
            err.get_error(),
            ResponseDataEnum::SingleValue(json!({}))
          )
        }
      }
    }
  }
}